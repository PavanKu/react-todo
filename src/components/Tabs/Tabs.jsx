import React, { useState, useEffect, useMemo } from "react";
import PropTypes from "prop-types";
import StyledTabGrp, { StyledTab } from "./Tabs.styled";
import {
  ARIA_ROLE_TABLIST,
  ARIA_ROLE_TAB,
} from "../../constants/ariaConstants";

export const DEFAULT_OPTIONS = ["All", "Active", "Completed"];

function Tabs({ active, options = DEFAULT_OPTIONS, handleClick }) {
  const [activeOption, setActiveOption] = useState(active);

  const handleTabClick = (value) => {
    if (handleClick && value !== activeOption) {
      handleClick(value);
    }
  };

  useEffect(() => {
    let validOption = active;
    const isValidOption = options.some((option) => option === validOption);
    if (!isValidOption) {
      validOption = options[0];
    }
    setActiveOption(validOption);
  }, [active, options]);

  const finalOptions = useMemo(() => {
    return options.filter((option) => option);
  }, [options]);

  return (
    <StyledTabGrp role={ARIA_ROLE_TABLIST}>
      {finalOptions.map((option) => (
        <StyledTab
          key={option}
          role={ARIA_ROLE_TAB}
          aria-selected={option === activeOption}
          onClick={() => handleTabClick(option)}
          isActive={option === activeOption}
        >
          {option}
        </StyledTab>
      ))}
    </StyledTabGrp>
  );
}

Tabs.propTypes = {
  active: PropTypes.string,
  options: PropTypes.arrayOf(PropTypes.string),
  handleClick: PropTypes.func,
};

export default Tabs;
