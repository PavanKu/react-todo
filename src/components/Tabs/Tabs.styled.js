import styled, { css } from "styled-components";

const StyledTabGrp = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  background: transparent;
`;

export const StyledTab = styled.button.attrs(() => ({
  type: "button",
}))`
  padding: 3px 7px;
  text-decoration: none;
  border: 1px solid transparent;
  border-radius: 3px;
  outline: none;
  background: transparent;
  cursor: pointer;
  margin-right: 6px;
  &:last-child {
    margin-right: 0;
  }

  ${(props) =>
    !props.isActive &&
    css`
      :hover {
        border-color: rgba(175, 47, 47, 0.1);
      }
    `}
  ${(props) =>
    props.isActive &&
    css`
      border-color: rgba(175, 47, 47, 0.2);
    `};
`;

export default StyledTabGrp;
