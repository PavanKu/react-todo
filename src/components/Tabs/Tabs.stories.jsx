import React from "react";
import { storiesOf } from "@storybook/react";
import { withKnobs, object, select } from "@storybook/addon-knobs";
import { action } from "@storybook/addon-actions";

import Tabs from "./Tabs";

const sampleOptions = ["Tab1", "Tab2", "Tab3"];
const activeOption = "Tab2";

const DefaultTabs = () => <Tabs />;
const TabsWithCustomActive = () => (
  <Tabs active={activeOption} options={sampleOptions} />
);
const Playground = () => (
  <Tabs
    active={select("active", sampleOptions)}
    options={object("options", sampleOptions)}
    handleClick={action("Tabs:Click:")}
  />
);

storiesOf("Tabs", module)
  .addDecorator(withKnobs)
  .add("Default Tab", DefaultTabs)
  .add("Custom Tab", TabsWithCustomActive)
  .add("Playground", Playground);
