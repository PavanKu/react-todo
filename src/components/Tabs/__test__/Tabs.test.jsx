import React from "react";
import { cleanup, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import Tabs, { DEFAULT_OPTIONS } from "../Tabs";
import {
  ARIA_ROLE_TAB,
  ARIA_TAB_SELECTED_ATTRIBUTE,
} from "../../../constants/ariaConstants";

describe("<Tab />", () => {
  afterEach(cleanup);

  it("should render without crash", () => {
    render(<Tabs />);
  });

  it("should render with default props", () => {
    const { getAllByRole } = render(<Tabs />);
    const tabElementsByRole = getAllByRole(ARIA_ROLE_TAB);
    const selectedTabElement = tabElementsByRole.filter(
      (element) => element.getAttribute(ARIA_TAB_SELECTED_ATTRIBUTE) === "true"
    )[0];

    expect(tabElementsByRole.length).toEqual(DEFAULT_OPTIONS.length);
    expect(selectedTabElement).toHaveTextContent(DEFAULT_OPTIONS[0]);
  });

  it("should render with props", () => {
    const sampleOptions = ["Tab1", "Tab2", "Tab3"];
    const activeOption = "Tab2";
    const { getAllByRole } = render(
      <Tabs options={sampleOptions} active={activeOption} />
    );
    const tabElementsByRole = getAllByRole(ARIA_ROLE_TAB);
    const selectedTabElement = tabElementsByRole.filter(
      (element) => element.getAttribute(ARIA_TAB_SELECTED_ATTRIBUTE) === "true"
    )[0];

    expect(tabElementsByRole.length).toEqual(sampleOptions.length);
    expect(selectedTabElement).toHaveTextContent(activeOption);
  });

  it("should invoke handlers", () => {
    let selectedTab;
    const selectedOption = DEFAULT_OPTIONS[1];
    const { getByText } = render(
      <Tabs handleClick={(tab) => (selectedTab = tab)} />
    );
    const secondtabElement = getByText(selectedOption);

    userEvent.click(secondtabElement);
    expect(selectedTab).toEqual(selectedOption);
  });

  it("should match snapshot", () => {
    const { container } = render(<Tabs />);
    expect(container).toMatchSnapshot();
  });
});
