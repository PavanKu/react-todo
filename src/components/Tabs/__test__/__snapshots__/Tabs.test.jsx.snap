// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tab /> should match snapshot 1`] = `
<div>
  <div
    class="sc-AxjAm dsugUK"
    role="tablist"
  >
    <button
      aria-selected="true"
      class="sc-AxirZ jNUtew"
      role="tab"
      type="button"
    >
      All
    </button>
    <button
      aria-selected="false"
      class="sc-AxirZ cCWkAh"
      role="tab"
      type="button"
    >
      Active
    </button>
    <button
      aria-selected="false"
      class="sc-AxirZ cCWkAh"
      role="tab"
      type="button"
    >
      Completed
    </button>
  </div>
</div>
`;
